### ISSUE TITLE
[Deprecation Info API] Provide enough meta data to allow Kibana Upgrade Assistant remove dynamic settings. 

### ISSUE THREAD

**Issue by jakelandis** [elastic/elasticsearch#83116]
The Deprecation Info API from Elasticsearch provides much of the source information for Kibana Upgrade assistant. For a very small subset of deprecated dynamic settings Kibana Upgrade Assistant can unset/remove these from the Elasticsearch cluster with a simple button click. A small subset is only supported today due to a lack of meta data needed to generalize the removal of these type of settings. In https://github.com/elastic/elasticsearch/issues/73089 the API gained support for generic meta data. 

This issue is to define the meta data needed to allow Kibana Upgrade Assistant to generalize the ability to remove/unset dynamic settings with a single button click. 

Since meta data is arbitrary key/value pairing... I would suggest something like

```json
...
  "_meta": {
    "actions" : {
      "remove_settings" : {
        "settings" : ["foo", "bar"]
    }
  }
...
```
This structure allows to add in additional future actions, and provides the context for that action. Since a single deprecation info message may have multiple settings that are deprecated, `settings` must be an array. In this case "foo" and "bar" would be the fully qualified setting name (e.g. `translog.retention.size`) and would be guaranteed to be able to be set/unset dynamically. Each deprecation message would have a remove_settings action iff the deprecation warning can be fully resolved by un-setting those settings. Some settings can be set both statically (elasticsearch.yml) and dynamically (via the API), the settings listed here would be guaranteed to have been set only dynamically (and thus can be unset dynamically). If a setting was actually set both statically and dynamically the deprecation message would not have the "remove_settings" entry since dynamically un-setting the setting would not resolve the deprecation warning. Further, there might be edge cases w.r.t. dependent settings or the scope of a deprecation warning..in all cases Elasticsearch must guarantee that no additional context/information is needed and un-setting the these setting listed dynamically will fully resolve the entire deprecation warning. 

When Kibana receives a deprecation message with the remove_settings action, it can be assured that taking that single action will resolve the full deprecation message. To remove/un-set the settings, Kibana will need to set the value to `null` for both persistent and transient settings. 

Alternatively, "actions" (or the like) could be a new top level item, reserving _meta for less structured information. However, I have a minor preference to keep it under _meta .

cc @alisonelizabeth @masseyke 

related: https://github.com/elastic/kibana/pull/123599

**Comment by alisonelizabeth** [comment]
This looks great. Thanks for opening up an issue around this @jakelandis!

I don't have a strong preference over keeping this information under `_meta` vs. adding a new top-level `actions` field. 

This may be a separate piece of work - but, there are other places where we're relying on the deprecation message in the ES API in order to determine UA actions. It'd be great if we could update these as well and create a unique action type. The two other actions that come to mind are: 

1 - Old indices that require a reindex - currently relying on the substring: "Index created before..."

```json
...
 "_meta": {
    "actions" : {
      "reindex" : {
        "indexName" : "foo" // I guess we already have access to the indexName so this may not be necessary
    }
  }
...
```

2 - Old ML snapshots - currently relying on the substring: "Model snapshot..."

```json
...
  "_meta": {
    "actions" : {
      "upgrade_ml_snapshots" : {
        "job_id":"deprecation_check_job",
         "snapshot_id":"1"
    }
  }
...
```

**Comment by cjcenizal** [comment]
Cross-linking to https://github.com/elastic/elasticsearch/issues/36324 for historical reference, since that's a similar idea we had awhile back.

**Comment by masseyke** [comment]
@jakelandis I think this will be pretty easy to do, but wanted to confirm before I go beyond the prototype stage. This is the format of a full response that we're looking for, right?
```
{
    "_meta": {
        "actions": {
            "remove_settings": {
                "settings": [
                    "discovery.zen.commit_timeout",
                    "discovery.zen.minimum_master_nodes"
                ]
            }
        }
    },
    "ccr_auto_followed_system_indices": [],
    "cluster_settings": [],
    "index_settings": {},
    "ml_settings": [],
    "node_settings": [
         {
            "details": "Remove the [discovery.zen.commit_timeout] setting. (nodes impacted: [keiths-mbp.lan])",
            "level": "critical",
            "message": "Setting [discovery.zen.commit_timeout] is deprecated",
            "resolve_during_rolling_upgrade": false,
            "url": "https://ela.st/es-deprecation-7-unused_zen_settings"
        },
        {
            "details": "Remove the [discovery.zen.minimum_master_nodes] setting. (nodes impacted: [keiths-mbp.lan])",
            "level": "critical",
            "message": "Setting [discovery.zen.minimum_master_nodes] is deprecated",
            "resolve_during_rolling_upgrade": false,
            "url": "https://ela.st/es-deprecation-7-unused_zen_settings"
        }
    ]
}

**Comment by dakrone** [comment]
I think we should step back a little bit and determine where we want the deprecation framework to go before adding something like this.

I definitely think this is something that we should work on adding, and that it would be quite useful, however, we're currently working on designing a new/enhanced deprecation framework, and is going to be part of that design. I don't want to end up with us working on a feasible design that works in a free-form field (the `_meta` field) in order to work in into our existing API without having a solid specification for it, as well as making sure it'll meet our future needs. Additionally, introducing something like this into the 7.17.x release feels a lot like a feature rushed in at the end rather than one well-considered.

I don't want to stop the discussion for this though, and I definitely think we want to head this way. But I also want to make sure we leave room in a design for things like automatic remediation, semi-automatic (one-click) remediation, and user-initiated remediation of deprecations.

**Comment by masseyke** [comment]
I started playing around with this and committed my partial work at https://github.com/elastic/elasticsearch/compare/7.17...masseyke:feature/deprecation-info-api-info-about-removable-settings?expand=1 so that I don't lose it. But I'll stop working on it now. In addition to Lee's concerns, I was also a little worried as I got into it that it's a fairly big change to the deprecation info API code. So there's risk of a regression.

**Comment by jakelandis** [comment]
> This is the format of a full response that we're looking for, right?

I meant for the _meta field to be 1:1 with the single deprecation issue. More like

```
 "node_settings": [
         {
            "details": "Remove the [discovery.zen.commit_timeout] setting. (nodes impacted: [keiths-mbp.lan])",
            "level": "critical",
            "message": "Setting [discovery.zen.commit_timeout] is deprecated",
            "resolve_during_rolling_upgrade": false,
            "url": "https://ela.st/es-deprecation-7-unused_zen_settings"
            "_meta": {
                 "actions": {
                    "remove_settings": {
                        "settings": [ "discovery.zen.commit_timeout"             ]
...
   
```

> I think we should step back a little bit and determine where we want the deprecation framework to go before adding something like this.

+1 let's discuss in person

**Comment by jpountz** [comment]
Thinking out loud, when the only reasonable action is to remove the setting, I wonder if we should bother users by showing it in the upgrade assistant at all, or if Elasticsearch should automatically take care of it?

**Comment by masseyke** [comment]
@dakrone you were going to add some comments about the format of the _meta block, right?

**Comment by dakrone** [comment]
> @dakrone you were going to add some comments about the format of the _meta block, right?

Yes, for this, I think we want something that does not use dynamic key names, and allows us to be flexible and not change BWC in the future, so rather than:

```json
"_meta": {"actions": {"remove_settings": {"settings": [...]}}}
```

It would be better to do something like:

```json
"_meta": {"actions": [{"action_type": "remove_settings", "objects": ["setting1", "setting2"]}]}
```

**Comment by cjcenizal** [comment]
@masseyke Is this closed by https://github.com/elastic/elasticsearch/pull/83601? Please reopen if I'm mistaken in closing this.

**Comment by masseyke** [comment]
Yeah it is -- it didn't auto-close because I merged it to 7.17. I'm working on porting it to master now (7.x and 8.x are so divergent at this point that a normal cherry-pick from master didn't make sense).
